### ISSUE TITLE
Canary Deployment Support

### ISSUE THREAD

**Issue by uMartinXu** [kubernetes-sigs/kernel-module-management#116]
### Issue Summary:
The Canary Deployment should be supported for user to execute some Testing to make sure the OOT driver can work properly on the cluster especially to prevent the cluster from being impacted by the new OOT driver modules insmod into the kernel on worker nodes.
### Suggested Priority(P1-P3) & Urgent(Urgent, medium, Low):
P2 & Medium
### Issue Detail:
It is quite possible that the Kernel on the work node can be impacted by some potential  issues bring by the OOT driver module. So KMMO directly deploy the driver container image to all the work nodes in the cluster without carefully testing on the clusters, it will cause huge problem to the whole cluster. KMMO reduce this risk by supporting the Canary Deployment. Using Canary Deployment, user can specify limited number of nodes( or specify some dedicated nodes) to deploy the driver container images. And after that, user can use some Canary testing (or run some real workloads) on these nodes and verify the stability and other quality of the driver. Only when the user has full confidence, user can then deploy the driver container image to all the other nodes in the clusters.
### Solution Proposal
User can specify the number of the nodes to deploy the driver, and KMMO can pick up the nodes for user to deploy the driver. 
User can also specify some dedicated nodes for KMMO to deploy the driver container image.
User can select to enable and disable Canary Deployment

**Comment by yevgeny-shnaidman** [comment]
It seems that canary mode can be implemented simply by labeling "canary" nodes and adding that label to the Selector field of the Module. Will that be a sufficient solution?

**Comment by GerrySeidman** [comment]
The approach I am using with SRO is just labeling nodes and deploying multiple SpecialResources, one per corresponding group of node-selected nodes.

I was assuming that the same technique wil work fine with KMMO.

**Comment by uMartinXu** [comment]
> It seems that canary mode can be implemented simply by labeling "canary" nodes and adding that label to the Selector field of the Module. Will that be a sufficient solution?

Yes, that is the basic logical. KMMO can label the proper nodes according to the user configure and the status of the node. For example select the nodes that has no relevant workload running on the old version of driver. Of course user can also label the group of nodes for KMMO to do the Canary deployment. as @GerrySeidman mentioned above. So at this case, KMMO just need to provide below primitive in CRD like
`canaryDeploymentNodes: DiverACanaryDeploymentLabel`

**Comment by yevgeny-shnaidman** [comment]
@uMartinXu what i meant is that customer can do node labeling and setting the Selector field himself, without any need for KMMO to do anything ( so no need to modify CRD or KMMO code). Also , letting KMMO decide which nodes to use for canary is not a good option: this decisions should be taken by customer only, because only he has all the data needed for making that decision

**Comment by uMartinXu** [comment]
@yevgeny-shnaidman using Module CRD should be a good idea for Canary deployment. And you are right it is not easy for KMMO to decide the nodes for Canary deployment. Let's have a try on it. 
BTW, will the selector also influence the building node select? Does the driver building process need to be the same selected nodes with deployment of Container driver?

**Comment by yevgeny-shnaidman** [comment]
@uMartinXu we don't really care where the build job is scheduled. It is containerized, so we nodes preference is not important

**Comment by qbarrand** [comment]
We use the `Module`'s main node selector for build placement: https://github.com/kubernetes-sigs/kernel-module-management/blob/5a772aa3496468ba9eaf1de4f28772d6902cdded/internal/build/job/maker.go#L114
This was implemented to solve issues in multi-arch clusters where we saw ARM nodes try to build x86 kmods or vice-versa. If that creates issues, I think we could have a dedicated node selector for builds. @uMartinXu if that's of interest to you, please open an issue 🙂

**Comment by chaitanya1731** [comment]
@qbarrand  we are quite interested in this feature and will file an issue soon.

**Comment by chaitanya1731** [comment]
Please refer - Separate node selector for build process #140
